File: Data_Structures/javascript/Stack/stack.js
// this is the Stack implementation using Arrays
class Stack {
    constructor() {
        this.data = [];
        this.top = 0;
    }

    push(element) {
      this.data[this.top] = element;
      this.top = this.top + 1;
    }

    length() {
      return this.top;
    }

    peek() {
        return this.data[this.top-1];
    }

    isEmpty() {
        return this.top === 0;
    }

    pop() {
        if( this.isEmpty() === false ) {
            this.top = this.top -1;
            return this.data.pop(); // removes the last element
        }
    }

    printStack() {
        var top = this.top - 1;
        while(top >= 0) {
            console.log(this.data[top]);
            top--;
        }
    }

    reverse() {
        this._reverse(this.top - 1 );
    }

    _reverse(index) {
        if(index != 0) {
            this._reverse(index-1);
        }
        console.log(this.data[index]);
    }
}

let stack = new Stack();

// Adding element to the stack
stack.push(10);
stack.push(20);
stack.push(30);

// Printing the stack element
// prints [10, 20, 30]
stack.printStack();

// returns 30
console.log('Peeking ', stack.peek());

console.log('Is it Empty: ', stack.isEmpty());

// returns 30 and remove it from stack
console.log('Popping', stack.pop());

// returns [10, 20]
stack.printStack();

File: Data_Structures/C/Stack/stack.c
// this is the stack implementation in c programming
#include <stdlib.h>
#include <stdio.h>
#include <limits.h>

struct Stack {
    int         top;
    int         capacity;
    int         *array;
};

struct Stack   *createStack(unsigned capacity)
{
    struct Stack *stack = malloc(sizeof(struct Stack));
    stack->capacity = capacity;
    stack->top = -1;
    stack->array = malloc(stack->capacity * sizeof(int));
    return stack;
}

int     isFull(struct Stack *stack)
{
    return stack->top == stack->capacity - 1;
}

void    push(struct Stack *stack, int data)
{
    if (isFull(stack))
        return;
    stack->top++;
    stack->array[stack->top] = data;
}

int     pop(struct Stack *stack)
{
    if (isFull(stack))
        return INT_MIN;
    return stack->array[stack->top--];
}

int     peek(struct Stack *stack)
{
    if (isFull(stack)) return INT_MIN;
    return stack->array[stack->top];
}

void    printStack(struct Stack *stack)
{
    for (int i = 0; i < stack->capacity; i++)
        printf("%d\n", stack->array[i]);
}

int     isempty(struct Stack *stack)
{
    return stack->top == -1;
}

int main()
{
    struct Stack* stack = createStack(100);

    push(stack, 10);
    push(stack, 20);
    push(stack, 30);
    push(stack, 40);
    push(stack, 50);
    push(stack, 60);
    push(stack, 90);

    printf("%d popped from stack\n", pop(stack));
    while(!isempty(stack)) {
        int data = pop(stack);
        printf("%d\n",data);
    }

    return 0;
}
